### DISCUSSION TITLE
unix: AF_UNIX socket send fd seems to be blocked

### DISCUSSION THREAD

**Discussion by number201724** [libuv/libuv#3837] [Q&A]
I hope to create a load balancer server and worker to cope with hundreds of thousands of links, worker connect to the load balancer server, then accept fd to continue execution, and manually control the number of workers (maybe there will be hot reload), but it seems that in *nix Under the platform, it prevents me from listening to UNIX domain socket.

https://github.com/libuv/libuv/blob/5102b2c093681dae2a90ea2196f868de78ec9957/src/unix/pipe.c#L100

e.g.:
server:
socket(AF_UNIX, ...)
bind("/var/tmp/xxxx.sock")
listen
accept()
....
sendmsg(fd)....

client:
socket(AF_UNIX)
connect("/var/tmp/xxxx.sock")
fd = recvmsg\()

**Comment by bnoordhuis** [comment]
I'm not sure what you're asking. Are you looking for the `uv_pipe_open()` function, to use an existing UNIX socket fd?

**Reply by number201724** [reply]
No, I want to use uv_pipe_init and use uv_pipe_bind to bind to unix socket after setting ipc=1. Then I use uv_write2 to send fd to other processes.

**Reply by bnoordhuis** [reply]
Okay. And what part of that isn't working?

Also, for the sake of preciseness: what does the fd in "send fd" refer to?

**Reply by number201724** [reply]
fd is socket fd, In the above code, it seems that ipc mode uv_pipe_t is prevented from using uv__pipe_listen

**Reply by number201724** [reply]
In windows, if the pipe is local, there is actually no difference, WSADuplicateSocket can send the socket correctly, just set ipc to true, and ignore the ipc check of the same function.

**Reply by number201724** [reply]
On *nix it seems that using AF_UNIX also correctly sends socket fd to other processes.

**Reply by bnoordhuis** [reply]
> seems that ipc mode uv_pipe_t is prevented from using uv__pipe_listen

Of course. Listen sockets can't recvmsg(), only accept().

**Reply by number201724** [reply]
> > seems that ipc mode uv_pipe_t is prevented from using uv__pipe_listen
> 
> Of course. Listen sockets can't recvmsg(), only accept().

I seem to understand, should it be like this: the server does not use ipc to create, and then uv_accept uses ipc to create a client?

**Reply by bnoordhuis** [reply]
Yes, that's right.
